File: sim-android/im_lib/src/main/java/com/sy/im/client/IMSEventListener.java
package com.sy.im.client;

import com.alibaba.fastjson.JSONObject;
import com.sy.im.interf.OnEventListener;
import com.sy.im.protobuf.MessageProtobuf;
import com.sy.im.ulit.MessageType;

import java.util.UUID;

public class IMSEventListener implements OnEventListener {

    private String userId;
    private String token;

    public IMSEventListener(String userId, String token) {
        this.userId = userId;
        this.token = token;
    }

    @Override
    public void dispatchMsg(MessageProtobuf.Msg msg) {
        System.out.println("方法消息给应用层IMSEventListener dispatchMsg："+msg);
    }

    @Override
    public boolean isNetworkAvailable() {
        return true;
    }

    @Override
    public int getReconnectInterval() {
        return 0;
    }

    @Override
    public int getConnectTimeout() {
        return 0;
    }

    @Override
    public MessageProtobuf.Msg getHandshakeMsg() {
        // 携带信息
        JSONObject jsonObj = new JSONObject();
        jsonObj.put("token", token);

        // 消息头部
        MessageProtobuf.Head.Builder headBuilder = MessageProtobuf.Head.newBuilder()
                .setMsgId(UUID.randomUUID().toString())
                .setMsgType(MessageType.HANDSHAKE.getMsgType())
                .setFromId(userId)
                .setExtend(jsonObj.toString())
                .setTimestamp(System.currentTimeMillis());
        // 构建消息
        MessageProtobuf.Msg.Builder builder = MessageProtobuf
                .Msg.newBuilder()
                .setHead(headBuilder.build());

        return builder.build();
    }

    @Override
    public MessageProtobuf.Msg getHeartbeatMsg() {
        MessageProtobuf.Head.Builder headBuilder = MessageProtobuf.Head.newBuilder()
                .setMsgId(UUID.randomUUID().toString())
                .setMsgType(MessageType.HEARTBEAT.getMsgType())
                .setFromId(userId)
                .setTimestamp(System.currentTimeMillis());
        // 构建消息
        MessageProtobuf.Msg.Builder builder = MessageProtobuf
                .Msg.newBuilder()
                .setHead(headBuilder.build());

        return builder.build();
    }

    @Override
    public int getServerSentReportMsgType() {
        return MessageType.SERVER_MSG_SENT_STATUS_REPORT.getMsgType();
    }

    @Override
    public int getClientReceivedReportMsgType() {
        return MessageType.CLIENT_MSG_RECEIVED_STATUS_REPORT.getMsgType();
    }

    @Override
    public int getResendCount() {
        return 0;
    }

    @Override
    public int getResendInterval() {
        return 0;
    }
}

File: sim-android/app/src/main/java/com/sy/im/client/IMSEventListener.kt
package com.sy.im.client

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.os.Build
import com.sy.im.interf.OnEventListener
import com.sy.im.protobuf.MessageProtobuf
import com.sy.im.provider.ContextProvider.context
import com.sy.im.message.MessageType
import com.sy.im.provider.AccountProvider
import com.sy.im.util.MessageProcessor
import java.util.*

class IMSEventListener : OnEventListener {
    override fun dispatchMsg(msg: MessageProtobuf.Msg?) {
        MessageProcessor.getInstance().receiveMsg(msg)
    }

    override fun isNetworkAvailable(): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val network = connectivityManager.activeNetwork
            val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
            networkCapabilities != null &&
                    (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                            networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR))
        } else {
            val networkInfo = connectivityManager.activeNetworkInfo
            networkInfo != null && networkInfo.isConnected
        }
    }


    override fun getHeartbeatMsg(): MessageProtobuf.Msg {
        val headBuilder = MessageProtobuf.Head.newBuilder()
            .setMsgId(UUID.randomUUID().toString())
            .setMsgType(MessageType.HEARTBEAT.msgType)
            .setFromId(AccountProvider.lastLoginUserId)
            .setTimestamp(System.currentTimeMillis())
        val builder = MessageProtobuf.Msg.newBuilder()
            .setHead(headBuilder.build())
        return builder.build()
    }

    // 0 为使用默认值
    override fun getReconnectInterval(): Int = 0

    override fun getConnectTimeout(): Int = 0

    override fun getForegroundHeartbeatInterval(): Int = 0

    override fun getBackgroundHeartbeatInterval(): Int = 0

    override fun getResendCount(): Int = 0

    override fun getResendInterval(): Int = 0
}